SimplenoteTests/NoticeContorllerTests.swift

```diff
@@ -1,33 +1,171 @@
+        XCTAssertEqual(notice.message, view.message)
+        XCTAssertEqual(notice.action?.title, view.actionTitle)
+        XCTAssertNotNil(view.handler)
     }
 
-    func testPerformanceExample() throws {
-        // This is an example of a performance test case.
-        self.measure {
-            // Put the code you want to measure the time of here.
+    func testMakeNoticeViewWithEmptyNotice() throws {
+        let action = NoticeAction(title: "") {
+            //No action
         }
+        let notice = Notice(message: "", action: action)
+
+        controller.present(notice)
+
+        let view = try XCTUnwrap(presenter.noticeView as? NoticeView)
+        XCTAssertEqual(notice.message, view.message)
+        XCTAssertNotEqual(notice.action?.title, view.actionTitle)
+        XCTAssertNotNil(view.handler)
     }
 
+    func testDismiss() {
+        let noticeA = Notice(message: "Message A", action: nil)
+
+        controller.present(noticeA)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+    }
+
+    func testDismissContinuesToNextNotice() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+        controller.present(noticeA)
+        controller.present(noticeB)
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.dismiss()
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+        XCTAssertEqual(presenter.actionLog, "presentNoticeView(Message A), dismissNotification(), presentNoticeView(Message B), dismissNotification(), ")
+    }
+}
+
+extension NoticeContorllerTests {
+    static var timeInterval = TimeInterval.zero
+    static var timerNoActionCompletionHandler: (Timer) -> Void = { (_) in
+        //no action
+    }
+}
+
+class MockNoticePresenter: NoticePresenter {
+    var presented: Bool = false
+    var dismissed: Bool = false
+    var listeningToKeyboard: Bool = false
+    var presentedViews = [NoticeView]()
+    var actionLog = String()
+
+    override func presentNoticeView(_ noticeView: NoticeView, completion: @escaping () -> Void) {
+        self.noticeView = noticeView
+        presentedViews.append(noticeView)
+        actionLog.append("presentNoticeView(\(noticeView.message ?? "")), ")
+        presented = true
+
+        completion()
+    }
+
+    override func dismissNotification(completion: @escaping () -> Void) {
+        dismissed = true
+        actionLog.append("dismissNotification(), ")
+        completion()
+    }
+
+    override func startListeningToKeyboardNotifications() {
+        listeningToKeyboard = true
+    }
+}
```

Comment:
If we use an action log there is no reason to keep also `presented` and `dismissed`. And instead of a list of presented views we can keep only the last one.
Mock presenter could look like this:
```
class MockNoticePresenter: NoticePresenter {
    enum Action: Equatable {
        case present(String?)
        case dismiss(String?)
    }

    var actionLog: [Action] = []

    var listeningToKeyboard: Bool = false
    var lastNoticeView: NoticeView?

    override func presentNoticeView(_ noticeView: NoticeView, completion: @escaping () -> Void) {
        lastNoticeView = noticeView
        actionLog.append(.present(noticeView.message))
        completion()
    }

    override func dismissNotification(completion: @escaping () -> Void) {
        let noticeView = try! XCTUnwrap(lastNoticeView)
        lastNoticeView = nil
        actionLog.append(.dismiss(noticeView.message))
        completion()
    }

    override func startListeningToKeyboardNotifications() {
        listeningToKeyboard = true
    }
}
```